### PR TITLE
Specify resource constraints for agent container

### LINKED ISSUE #173: Allow easy configuration of compute resources
**Is your feature request related to a problem? Please describe.**
Every pipeline has different needs regarding CPU and memory resources. Jenkinsfile authors should be able to easily adapt to their needs. Further, good defaults should be provided depending on the used language/framework.

### PR DESCRIPTION
**Pull request by michaelsauter**
Closes #173.

### CODE REVIEW

**Comment by michaelsauter** [comment]
Replaces #151.

**Comment by michaelsauter** [comment]
Marked as "breaking change" as the property `podAlwaysPullImage` changes to `alwaysPullImage` to better reflect that this is a container setting, not a pod setting. Impact is going to be low as this is usually not set explicitly and I expect only very few people to use this.

**Comment by michaelsauter** [comment]
FYI @SimonGolms

File: docs/modules/ROOT/pages/index.adoc [hunk 1]
```diff
@@ -165,18 +165,30 @@ When you write stages, you have access to both global variables (defined without
 | image
 | Container image to use for the Jenkins agent container. This value is not used when "podContainers" is set.
 
+| alwaysPullImage
```
Comment (henrjk) [inline docs/modules/ROOT/pages/index.adoc]: The default should be mentioned here as well.

File: docs/modules/ROOT/pages/index.adoc [hunk 2]
```diff
@@ -165,18 +165,30 @@ When you write stages, you have access to both global variables (defined without
...
+
+| resourceLimitCpu
+| Maximum CPU the container can use - defaults to 300mi. This value is not used when "podContainers" is set.
+
 | podLabel
 | Pod label, set by default to a random label to avoid caching issues. Set to a stable label if you want to reuse pods across builds.
 
 | podContainers
-| Custom pod containers to use. By default, only one container is used, and it is configure automatically. If you need to run multiple containers (e.g. app and database), then you can configure the containers via this property.
+| Custom pod containers to use if the default, automatically configured container is not suitable for your use case (e.g. if you need multiple containers such as app and database). Please see https://github.com/jenkinsci/kubernetes-plugin#pod-and-container-template-configuration for more information.
```
Comment (henrjk) [inline docs/modules/ROOT/pages/index.adoc]: I like there is a link to the docs.

**Comment by michaelsauter** [comment]
@henrjk Thanks for the review. I've added changes based on your feedback in c10f0d1. I left the example with the `podContainers` configuration because it shows how to control the container completely, not just the resource constraints. But I made the wording a bit clearer (hopefully).

**Comment by michaelsauter** [comment]
@henrjk Are the changes fine with you?

I've also updated the defaults to higher settings after giving those values some test runs. It's unfortunate that these values are quite high but I am afraid we can't lower the defaults without creating lots of problems in the next release. For new components, I hope that we get lower defaults for some quickstarters into the 3.0 release.

**Comment by henrjk** [comment]
Yes the changes are fine with me.

**Comment by michaelsauter** [comment]
@henrjk Cool - please approve again. Since I have added a new commit, your previous approval was dismissed so merging is still blocked.

**Pull request merged**